Design Proposal for Etcd OperatorTask Operator. (#3)
seshachalam-yv/etcd-druid: docs/development/etcdoperatortask-design.md

# EtcdOperatorTask Controller Implementation Design

> This document builds upon the initial proposal outlined in [05-etcd-operator-tasks](https://github.com/gardener/etcd-druid/blob/main/docs/proposals/05-etcd-operator-tasks.md) and describes the design and implementation of the **EtcdOperatorTask** controller. The controller enables out-of-band operational tasks for etcd clusters managed by etcd-druid. The design emphasizes extensibility, clear separation of concerns, and robust status tracking.

Reviewer: ```suggestion
This document builds upon the initial proposal outlined in [05-etcd-operator-tasks](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/05-etcd-operator-tasks.md) and describes the design and implementation of the **EtcdOperatorTask** controller. The controller enables out-of-band operational tasks for etcd clusters managed by etcd-druid. The design emphasizes extensibility, clear separation of concerns, and robust status tracking.
```

- The `spec` is immutable (`kubebuilder:validation:Immutable`) to ensure task intent cannot be changed after creation.
- The `config` field uses `runtime.RawExtension` to flexibly support task-specific parameters, validated via admission webhook.
- `ownerEtcdReference` under the spec field is renamed to `etcdReference` for clarity.
- The `LastOperation` struct now uses a `Description` field instead of `Name` and `Reason` for better expressiveness.

### Go API Definition

```go
// +kubebuilder:object:root=true
// +kubebuilder:resource:path=etcdoperatortasks,shortName=eot;eots,scope=Namespaced
// +kubebuilder:subresource:status
type EtcdOperatorTask struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    Spec   EtcdOperatorTaskSpec   `json:"spec"`
    Status EtcdOperatorTaskStatus `json:"status,omitempty"`
}

// +kubebuilder:validation:XPreserveUnknownFields
// +kubebuilder:validation:Immutable
type EtcdOperatorTaskSpec struct {
    // +required
    Type v1alpha1.EtcdOperatorTaskType `json:"type"`

    // Config is task-specific key/value parameters.
    // +required

>     Config *Config `json:"config,omitempty"` 

Reviewer: ```suggestion
    Config *runtime.RawExtension `json:"config,omitempty"` 
```
Shouldn't this be of this type as suggested in the modification list above.

Reviewer 2: The initial plan was to go forward with `runtime.RawExtension`, but during the discussion, @renormalize suggested going forward with [this approach](https://github.com/kubernetes/kubernetes/blob/82db38a23c7820b1924d89f458fd368023f3980c/staging/src/k8s.io/api/core/v1/types.go#L36) since we'll avoid dealing with unmarshalling of the config field.
Will do the needful modifications.

    // TTLSecondsAfterFinished controls how long the status+pod stays around.
    // +optional
    // +kubebuilder:validation:XValidation:message="TTLSecondsAfterFinished must be greater than 0",rule="!has(self) || self > 0"
    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`

    // +optional

>     EtcdReference types.NamespacedName `json:"etcdReference,omitempty"`

Reviewer: Can you also describe in what situation will this be empty? as in the proposal its wan't marked as optional.

Reviewer 2: In case of on-demand operations such as `etcd-copybackup-tasks`, we don't need an etcd reference since it only depends on the backup buckets. Since these operations are to be run via the operator task, we decided to keep this as an optional field.

}

type Config struct {
    OnDemandSnapshotConfig *OnDemandSnapshotConfig

>     // Add config fields for other tasks

Reviewer: ```suggestion
    // Configs for more supported tasks shall be added here subsequently 
```
Also, just for my lack of implmentation details, shouldn't these also have json tag with omitEmpty?

  Phase OperationPhase `json:"phase"`
  // LastTransitionTime records the timestamp of the most recent state transition, marking when the operation moved from its previous state to the current value specified in the 'State' field.
  LastTransitionTime *metav1.Time `json:"lastTransitionTime"`
  // A human readable message indicating details about the last operation.
  Description string `json:"description"`
}

// LastError stores details of the most recent error encountered for the task.
type LastError struct {
  // Code is an error code that uniquely identifies an error.
  Code ErrorCode `json:"code"`
  // Description is a human-readable message indicating details of the error.
  Description string `json:"description"`
  // ObservedAt is the time at which the error was observed.
  ObservedAt *metav1.Time `json:"observedAt"`
}

// TaskState represents the state of the task ie the status of the CR.
type TaskState string

const (
    TaskStatePending TaskState = "Pending"
    TaskStateRejected TaskState = "Rejected"
    TaskStateInProgress TaskState = "InProgress"
    TaskStateFailed TaskState = "Failed"
    TaskStateSucceeded TaskState = "Succeeded"

> )

Reviewer: The implementation behavior below assumes the state transitions are strict across tasks. 
Should there be a state diagram here to ensure the implementor of future tasks understand what state a task can go so that they abide by the flow and map the behavior accordingly in their implementation.

// OperationState represents the state of last operation run via the task.
type OperationState string

const (
    OperationStateInProgress OperationState = "InProgress"
    OperationStateCompleted OperationState = "Completed"
    OperationStateFailed OperationState = "Failed"
)

type OperationPhase string
const (
    OperationPhaseAdmit OperationPhase = "Admit"
    OperationPhaseRunning OperationPhase = "Run"
    OperationPhaseCleanup OperationPhase = "Cleanup"
)
```

> </details>

Reviewer: ```suggestion
```
This should not be there right?

---

### Example: EtcdOperatorTask YAML

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: EtcdOperatorTask
metadata:
    name: <name of operator task resource>
    namespace: <namespace>
    generation: <specific generation of the desired state>
spec:
    type: <type/category of supported out-of-band task>
    ttlSecondsAfterFinished: <time-to-live to garbage collect the custom resource after it has been completed>
    config: <task specific configuration>
    etcdReference: <refer to corresponding etcd owner name and namespace for which task has been invoked>
status:
    observedGeneration: <specific observedGeneration of the resource>
    state: <last known current state of the out-of-band task>
    initiatedAt: <time at which task move to any other state from "pending" state>
    lastErrors:
    - code: <error-code>
      description: <description of the error>
      observedAt: <time the error was observed>
    lastOperation:
      description: <meaningful description>
      state: <task state as seen at the completion of last operation>
      lastTransitionTime: <time of transition to this state>

```

> </details>

Reviewer: ```suggestion
```
Same as above I believe.

```

**Interface responsibilities:**

- `Admit`: One-time precondition check before task execution.
- `Run`: Main execution logic, called on each reconcile until completion.
- `Cleanup`: Post-completion or TTL-based cleanup.
- `Name`, `Type`, `EtcdReference`: Used for metrics and logging.

## TaskHandler Instantiation

The controller uses a factory function (`createTaskHandlerInstance`) to instantiate the correct handler for each task type. This enables easy extension and decouples task logic from the reconciler.

```go
// Add a new TaskHandler as shown below:
func (r *Reconciler) createTaskHandlerInstance(task *v1alpha1.EtcdOperatorTask) (tasks.TaskHandler, error) {
    switch task.Spec.Type {
    case v1alpha1.EtcdOperatorTaskTypeOnDemandSnapshot:
        return ondemandsnapshot.New(r.client, r.logger, task) // New() initializes the TaskHandler.
    // Add more cases for other task types
    default:
        return nil, fmt.Errorf("unsupported task type: %s", task.Spec.Type)
    }
}

```

### Adding New Task Types

To add a new task type:
1. Implement the `TaskHandler` interface for your task.
2. Register the handler in `createTaskHandlerInstance` with a new case for your type.

> 

Reviewer: How do you handle RBAC for tasks? Should there be granularity on who can execute what kind of operation task or we have a blanket RBAC for all tasks. 

Also shouldn't we audit trail any such task execution? I believe, we should have it included in the implementation where an execution to task handler is audit trailed?

- **TTL Validation:** If `ttlSecondsAfterFinished` is set, ensures it is greater than zero.

## Reconciliation Flow

The controller's reconciliation loop manages the lifecycle of each `EtcdOperatorTask` resource, including validation, execution, status updates, and cleanup.

```go
func (r *TaskReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
    // The below block checks for the OperatorTask.
    task := &v1alpha1.EtcdOperatorTask{}
    if err := r.client.Get(ctx, req.NamespacedName, task); err != nil {
        if client.IgnoreNotFound(err) != nil {
            return reconcile.Result{}, err
        }
        return reconcile.Result{}, nil
    }

    logger := r.logger.WithValues("runId", string(controller.ReconcileIDFromContext(ctx)))
    // create a instance of TaskHandler
    taskHandlerInstance := r.createTaskHandlerInstance(task)
    
    // Triggers the deletion flow in case the task is in a completed state or if it has been marked for deletion.

>     if task.IsCompleted() || task.IsMarkedForDeletion() {

Reviewer: task.IsMarkedForDeletion --> I don't see any corresponding state for this ?

Reviewer 2: Since this is a high level implementation, `isMarkedForDeletion()` is basically a check to see if the deletion timestamp is added to the resource. This can arise in case of forced-deletion where the operator deletes the CR.

func New(k8sclient client.Client, logger logr.Logger, task *v1alpha1.EtcdOperatorTask) (operatortask.OperatorTask, error) {
     
    config,err := decodeOnDemandSnapshotConfig(task.Spec.Config)
    if err != nil {
        return nil, druiderr.WrapError(err, operatortask.ERR_INVALID_CONFIG, operationNew, "failed to decode config for OnDemandSnapshot")
    }
    
    return &OnDemandSnapshot{
        client:        k8sclient,
        logger:        logger,
        name:          task.Name,
        httpClient:    &http.Client{Timeout: config.Timeout},
        etcdReference: types.NamespacedName(*task.Spec.EtcdRef),
        config:        config,
    }, nil
}

func (o *OnDemandSnapshot) Run(ctx context.Context) *operatortask.TaskResult {
    // Implement the snapshot logic
    etcd := &v1alpha1.Etcd{}
    if err := o.client.Get(ctx, o.etcdReference, etcd); err != nil {
        return &operatortask.TaskResult{Description: "Failed to get Etcd resource", Error: err}
    }

    snapType := "full"
    if o.config.SnapshotType != nil && *o.config.SnapshotType != "" {
        snapType = *o.config.SnapshotType
    }

>     o.logger.Info("Snapshot config", "snapshotType", snapType, "timeout", o.config.Timeout)

Reviewer: At what level will the logs be visible for operator task, Druid Plutono dashbord in seed or Shoot Plutono dashboard?

This document builds upon the initial proposal outlined in [05-etcd-operator-tasks](https://github.com/gardener/etcd-druid/blob/main/docs/proposals/05-etcd-operator-tasks.md) and describes the design and implementation of the **EtcdOperatorTask** controller. The controller enables out-of-band operational tasks for etcd clusters managed by etcd-druid. The design emphasizes extensibility, clear separation of concerns, and robust status tracking.

## Overview

The `EtcdOperatorTask` custom resource (CR) provides a generic mechanism to trigger and manage operational tasks (such as on-demand snapshots, maintenance, etc.) for etcd clusters. Each task is represented as a CR instance, with its lifecycle managed by a single controller. This approach decouples operational logic from the core reconciliation loop and ensures seamless extensibility for introducing new task types.

## Custom Resource API Design

The `EtcdOperatorTask` CRD is defined under the `v1alpha1` API version. It is subject to change and will follow the [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).

**Key design decisions:**

> The authors slightly modified CRD for implementation purposes introduced to handle out-of-band tasks as initially proposed by [05-etcd-operator-tasks](https://github.com/gardener/etcd-druid/blob/main/docs/proposals/05-etcd-operator-tasks.md)

Reviewer 3: ```suggestion
 The authors have made minor modifications to the CRD for implementation purposes, in order to support out-of-band tasks as originally proposed in [05-etcd-operator-tasks](https://github.com/gardener/etcd-druid/blob/main/docs/proposals/05-etcd-operator-tasks.md)
```

The wording was slightly confusing to read, can you change it to this?

    // +required
    Config runtime.RawExtension `json:"config,omitempty"` 

    // TTLSecondsAfterFinished controls how long the status+pod stays around.
    // +optional
    // +kubebuilder:validation:XValidation:message="TTLSecondsAfterFinished must be greater than 0",rule="!has(self) || self > 0"
    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`

    // +optional
    EtcdReference types.NamespacedName `json:"etcdReference,omitempty"`

}
```

> [!Note]
> Use `runtime.RawExtension` for the `config` field to enable each task type to define its own schema dynamically. Ensure this field is validated through an admission webhook to enforce type safety and provide immediate feedback to users.

### Status Subresource

The `status` field tracks the progress and outcome of each task, including state transitions, errors, and operation history.

```go
// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
type EtcdOperatorTaskStatus struct {
  // State is the last known state of the task.
  State TaskState `json:"state"`
  // Time at which the task has moved from "pending" state to any other state.
  // +optional

>   InitiatedAt *metav1.Time `json:"initiatedAt"`

Reviewer 3: ```suggestion
  InitiatedAt *metav1.Time `json:"initiatedAt,omitempty"`
```

>   // LastError represents the errors when processing the task. Will have a limit of 10 entries at a time.

Reviewer 3: ```suggestion
  // LastErrors represents the errors when processing the task. Will have a limit of 10 entries at a time.
```

  // +optional
  LastErrors []LastError `json:"lastErrors,omitempty"`
  // Captures the last operation status if task involves many stages.
  // +optional
  LastOperation *LastOperation `json:"lastOperation,omitempty"`
}

type LastOperation struct {
  // Status of the last operation, one of pending, progress, completed, failed.
  State OperationState `json:"state"`

>   // LastTransitionTime is the time at which the operation state last transitioned from one state to another.

Reviewer 3: ```suggestion
  // LastTransitionTime is the time at which the operation state last transitioned from one previous state to current state i.e 'State'
```

Maybe not exactly the same as I suggested, but can you please make it more concrete here, because when someone reads this, they're already reading about the `LastOperation` struct which is a an atomic unit that encompasses the `State`, so we can just mention that the `lastTransitionTime` captures the time when the operation state transitioned from previous state to current `State` that the struct is in, i.e `lastOperation.state`. Correct me if my understanding is wrong.

# EtcdOperatorTask Controller Design

This document describes the design and implementation of the **EtcdOperatorTask** controller, which enables out-of-band operational tasks for etcd clusters managed by etcd-druid. The design focuses on extensibility, clear separation of concerns, and robust status tracking.

> ---

Reviewer 4: Do we really need the page separators throughout the doc? Personally, it seems unnecessary to me, and disrupts the reading flow.

Author: Thank you for pointing this out. I have removed the page separators to improve readability.

## Overview

> The `EtcdOperatorTask` custom resource (CR) provides a generic mechanism to trigger and manage operational tasks (such as on-demand snapshots, maintenance, etc.) for etcd clusters. Each task is represented as a CR instance, with its lifecycle managed by a dedicated controller. This approach decouples operational logic from the core reconciliation loop and enables easy extension for new task types.

Reviewer 4: > Each task is represented as a CR instance, with its lifecycle managed by a dedicated controller

To me, this sentence seems to imply that each CR instance has its own dedicated controller. But in the context of this proposal, it seems to read like "each operation type gets its own controller", which I don't think is true, correct?

Please re-word this sentence to remove this ambiguity.

Author: I have clarified the sentence to indicate that a single controller manages the lifecycle of all CR instances.

---

## Custom Resource API Design

The `EtcdOperatorTask` CRD is defined under the `v1alpha1` API version. It is subject to change and will follow the [Kubernetes Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).

**Key design decisions:**

- The `spec` is immutable (`kubebuilder:validation:Immutable`) to ensure task intent cannot be changed after creation.
- The `config` field uses `runtime.RawExtension` to flexibly support task-specific parameters, validated via admission webhook.
- `ownerEtcdReference` under the spec field is renamed to `etcdReference` for clarity.

> - The `LastOperation` struct now uses a `Description` field instead of `Name` and `Reason` for better expressiveness.

Reviewer 4: I see that these points have been written `in addition` to an already defined API, but I don't see that API being shown or referenced anywhere. Can you please link to the original proposed API in the DEP? That would give context to readers about what was the original API, and what changes you have proposed here.

Author: Added reference

---

### Go API Definition

<details>

> <summary>Show EtcdOperatorTask CRD Go Definition</summary>

Reviewer 4: This need not be part of a `details` section, but rather part of the main page flow itself. Reason being, the API is an important part of your document, and it doesn't make sense to keep it hidden by default. 

Same applies for the other hidden `section` fields in the document that show the API and example CR.

Author: Removed the details section.

    Config runtime.RawExtension `json:"config,omitempty"` 

    // TTLSecondsAfterFinished controls how long the status+pod stays around.
    // +optional
    // +kubebuilder:validation:XValidation:message="TTLSecondsAfterFinished must be greater than 0",rule="!has(self) || self > 0"
    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`

    // +optional
    EtcdReference types.NamespacedName `json:"etcdReference,omitempty"`

}

```
</details>

> **Best Practice:**
> Use `runtime.RawExtension` for the `config` field to allow each task type to define its own schema. Validate this field with an admission webhook for type safety and user feedback.

---

### Status Subresource

The `status` field tracks the progress and outcome of each task, including state transitions, errors, and operation history.

<details>
<summary>Show EtcdOperatorTaskStatus Go Definition</summary>

```go
// EtcdOperatorTaskStatus is the status for a EtcdOperatorTask resource.
type EtcdOperatorTaskStatus struct {
  // State is the last known state of the task.
  State TaskState `json:"state"`
  // Time at which the task has moved from "pending" state to any other state.
  // +optional

>   InitiatedAt *metav1.Time `json:"initiatedAt"`

Reviewer 4: If `Status.State` is `Rejected`, then would `initiatedAt` still record the time at which it moved from `Pending` to `Rejected`? If yes, then the name seems non-intuitive, since the task was not really initiated, since it was rejected.

Author: As we discussed out of band discussion, The `status.initiatedAt` field will be set when the task transitions from `Pending` to `InProgress`.

  // LastError represents the errors when processing the task. Will have a limit of 10 entries at a time.
  // +optional
  LastErrors []LastError `json:"lastErrors,omitempty"`
  // Captures the last operation status if task involves many stages.
  // +optional
  LastOperation *LastOperation `json:"lastOperation,omitempty"`
}

type LastOperation struct {
  // Status of the last operation, one of pending, progress, completed, failed.

>   State OperationState `json:"state"`

Reviewer 4: How is this different from `Status.State`? I see an overlap of information here.

  // LastTransitionTime is the time at which the operation state last transitioned from one state to another.

>   LastTransitionTime *metav1.Time `json:"lastTransitionTime"`

Reviewer 4: Same as above. Since this already denotes the last transition time for the operation, do we really need `Status.InitiatedAt`?

Reviewer 2: `status.initiatedAt` is for recording when the operation moves from pending to any other state (rejected/in-progress), whereas the `status.lastOperation.lastTransitionTime` is to record the time when the operation transitions from the various operation states such as `Admit`, `In-Progress` and `Cleanup`.

type LastError struct {
  // Code is an error code that uniquely identifies an error.
  Code ErrorCode `json:"code"`
  // Description is a human-readable message indicating details of the error.
  Description string `json:"description"`
  // ObservedAt is the time at which the error was observed.
  ObservedAt *metav1.Time `json:"observedAt"`
}

// TaskState represents the state of the task ie the status of the CR.
type TaskState string

const (
  TaskStateFailed TaskState = "Failed"
  TaskStatePending TaskState = "Pending"
  TaskStateRejected TaskState = "Rejected"
  TaskStateSucceeded TaskState = "Succeeded"
  TaskStateInProgress TaskState = "InProgress"
)

// OperationState represents the state of last operation run via the task.
type OperationState string

const (
  OperationStateInProgress OperationState = "InProgress"
  OperationStateCompleted OperationState = "Completed"
  OperationStateFailed OperationState = "Failed"
)

```
</details>

---

### Example: EtcdOperatorTask YAML

<details>
<summary>Show Example EtcdOperatorTask YAML</summary>

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: EtcdOperatorTask
metadata:
    name: <name of operator task resource>

>     namespace: <cluster namespace>

Reviewer 4: ```suggestion
    namespace: <namespace>
```

Is it necessary that the `EtcdOperatorTask` be run in the same namespace as the referenced `Etcd` resource?

    }

    for _, step := range reconcileStepFns {
        ctx.Logger.Info("Executing step", "step", step)
        result := step(ctx, taskObjKey, taskHandler)
        if ctrlutils.ShortCircuitReconcileFlow(result) {
            return result
        }
    }

    return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
}
```
</details>

---

### Deletion Flow

<details>
<summary>Show Task Deletion Flow</summary>

```go
func (r *Reconciler) triggerTaskDeletionFlow(
    ctx context.Context,
    logger logr.Logger,
    taskObjKey client.ObjectKey,
    taskHandler tasks.TaskHandler,
) ctrlutils.ReconcileStepResult {
    if task.IsCompleted() && !task.IsMarkedForDeletion() {
        if !task.TTLHasExpired() {
            return ctrlutils.ReconcileAfter(task.Spec.TTLSecondsAfterFinished, "Task completed, waiting for TTL to expire")
        }
    }

    deletionStepFns := []reconcileFn{
        r.recordTaskDeletionStartOperation, // update the status. Skip if already done.
        r.cleanupTaskResources, // Cleanup of any kind of resources used to run the task. Runs the interface method.

>         r.recordTaskDeletionSuccessOperation,

Reviewer 4: Might not be necessary, since the resource is about to be deleted anyway. We can save an additional status update call here.

Author: Removed

    // No-op for snapshot tasks
    return &operatortask.TaskResult{Description: "Cleanup done", Completed: true}
}

```
</details>

#### Example YAML for an On-Demand Snapshot Task

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: EtcdOperatorTask
metadata:
  name: on-demand-snapshot-task
  namespace: default
spec:
  type: OnDemandSnapshot
  etcdRef:
    name: etcd-test
    namespace: default
  config:
    snapshotType: "full"
    timeoutSeconds: 60
  ttlSecondsAfterFinished: 600
```

### Registering a New OperatorTask

1. Implement the `TaskHandler` and its constructor.
2. Add the new handler to `createTaskHandlerInstance`:
   ```go
   func (r *Reconciler) createTaskHandlerInstance(task *v1alpha1.EtcdOperatorTask) (tasks.TaskHandler) {
       switch task.Spec.Type {
       case v1alpha1.<NewTaskType>:
           return <NewTaskType>.New(r.client, r.logger, task)
       // Add more cases for other task types
       default:
       }
   }
   ```

**Benefits:**
- **Extensible:** New OperatorTasks can be added without modifying the core controller logic.
- **Decoupled:** OperatorTask implementations are independent from the reconciler logic.

---

## Next Steps

> - Design and implement HTTP endpoints in the etcd-backup-restore server to expose the etcd maintenance API, supporting both sync and async operations.

Reviewer 4: When you say async, you're talking about doing it via an API resource that etcdbr reads and then performs the snapshot upload and record the status result to?

Author: Current snapshot API is not fully RESTful and does not support async operations. As part of this design, we plan to refactor the snapshot API to follow REST principles and support both synchronous and asynchronous workflows. The goal is to provide standard HTTP(S) endpoints to trigger operations and poll their status, making it easier to integrate and automate etcd maintenance tasks. The exact API shape is yet to be designed.
